File: frontend/src/types/widget.ts
export type WidgetAlignment = "left" | "center" | "right";
export type WidgetBackgroundType = "solid" | "gradient" | "image";
export type WidgetAggregationMethod = "sum" | "mean" | "median" | "count" | "percentile";

export type SummaryMetricId =
  | "count"
  | "sum"
  | "average"
  | "median"
  | "mode"
  | "minimum"
  | "maximum"
  | "range"
  | "standard_deviation"
  | "variance"
  | "percentile"
  | "growth_rate"
  | "moving_average"
  | "distribution_skew"
  | "top_values";

export type SupportedWidgetType =
  | "bar"
  | "line"
  | "area"
  | "pie"
  | "donut"
  | "scatter"
  | "heatmap"
  | "radar"
  | "histogram"
  | "stacked_bar"
  | "table"
  | "kpi"
  | "text"
  | "filter";

export interface WidgetStyleConfig {
  background_type?: WidgetBackgroundType;
  background_color?: string;
  gradient_from?: string;
  gradient_to?: string;
  background_image?: string;
  font_size?: number;
  font_weight?: number;
  bold?: boolean;
  italic?: boolean;
  border_color?: string;
  border_width?: number;
  border_radius?: number;
  highlight_border?: boolean;
  card_shadow?: string;
  hover_shadow?: string;
  padding?: number;
  spacing?: number;
  alignment?: WidgetAlignment;
}

export interface WidgetMetricConfig {
  field?: string;
  aggregation?: WidgetAggregationMethod;
  percentile?: number;
  moving_average_window?: number;
  top_values_limit?: number;
  show_in_header?: boolean;
  visible_metrics?: SummaryMetricId[];
}

export interface ChartConfig {
  x_field: string;
  y_fields: string[];
  group_field?: string;
  aggregation: string;
  colors: string[];
  stacked: boolean;
  show_values: boolean;
  orientation: string;
  x_axis_label?: string;
  y_axis_label?: string;
  metric_name?: string;
  card_description?: string;
  show_legend?: boolean;
  show_tooltip?: boolean;
  show_grid?: boolean;
  showSummaryMetrics?: boolean;
  histogram_bins?: number;
  style_config?: WidgetStyleConfig;
  metric_config?: WidgetMetricConfig;
  /** Allow extra properties used by KPI, text, and filter widgets */
  [key: string]: unknown;
}

export interface LayoutPosition {
  x: number;
  y: number;
  w: number;
  h: number;
  min_w?: number;
  min_h?: number;
  position?: number;
}

export interface Widget {
  id: string;
  dashboard_id: string;
  connection_id?: string | null;
  type: SupportedWidgetType | string;
  title?: string;
  prompt_used?: string;
  query_config?: Record<string, unknown>;
  chart_config: ChartConfig;
  layout_position: LayoutPosition;
  data?: Record<string, unknown>[];
  cached_data?: Record<string, unknown>[];
  created_at: string;
}

export interface WidgetCreate {
  dashboard_id: string;
  type: string;
  title?: string;
  connection_id?: string;
  query_config: Record<string, unknown>;
  chart_config: Record<string, unknown>;
  layout_position: Record<string, unknown>;
}

export interface WidgetUpdate {
  title?: string;
  type?: string;
  query_config?: Record<string, unknown>;
  chart_config?: Record<string, unknown>;
  layout_position?: Record<string, unknown>;
}

File: backend/schemas/widget.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ChartConfig(BaseModel):
    x_field: str
    y_fields: list[str]
    group_field: str | None = None
    aggregation: str = "sum"
    colors: list[str] = Field(default_factory=list)
    stacked: bool = False
    show_values: bool = True
    orientation: str = "vertical"
    x_axis_label: str | None = None
    y_axis_label: str | None = None
    metric_name: str | None = None
    card_description: str | None = None
    prefix: str | None = None
    suffix: str | None = None
    show_legend: bool = True
    show_tooltip: bool = True
    show_grid: bool = True
    histogram_bins: int | None = None
    style_config: dict[str, Any] = Field(default_factory=dict)
    metric_config: dict[str, Any] = Field(default_factory=dict)

    model_config = {"extra": "allow"}


class LayoutPosition(BaseModel):
    x: int
    y: int
    w: int
    h: int
    min_w: int = 2
    min_h: int = 2
    position: int | None = None


class WidgetCreate(BaseModel):
    dashboard_id: str
    type: str
    title: str | None = None
    connection_id: str | None = None
    query_config: dict
    chart_config: dict
    layout_position: dict


class WidgetRead(BaseModel):
    id: uuid.UUID
    dashboard_id: uuid.UUID
    connection_id: uuid.UUID | None = None
    type: str
    title: str | None = None
    prompt_used: str | None = None
    query_config: dict = Field(default_factory=dict)
    chart_config: dict
    layout_position: dict
    data: list[dict[str, Any]] | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class WidgetUpdate(BaseModel):
    title: str | None = None
    type: str | None = None
    query_config: dict | None = None
    chart_config: dict | None = None
    layout_position: dict | None = None
